format/private/format.sh

```diff
@@ -123,160 +128,79 @@ case "$mode" in
+function run-lints {
+  local lang="$1" && shift
+  local fmtname="$1" && shift
+  local bin="$1" && shift
+  local args="$1" && shift
+  local tuser
+  local tsys
+
+  local files=$(ls-files "$lang" $@)
+  if [ -n "$files" ] && [ -n "$bin" ]; then
+    echo "Formatting ${lang} with ${fmtname}..."
+    case "$lang" in
+    'Protocol Buffer')
+        ( for file in $files; do
+          "$bin" $args $file >&2
+        done ; times ) | ( read _ _; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+        ;;
+      Go)
+        # gofmt doesn't produce non-zero exit code so we must check for non-empty output
+        # https://github.com/golang/go/issues/24230
+        if [ "$mode" == "check" ]; then
+          GOFMT_OUT=$(mktemp)
+          (echo "$files" | tr \\n \\0 | xargs -0 "$bin" $args > "$GOFMT_OUT" ; times ) | ( read _ _; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+          NEED_FMT="$(cat $GOFMT_OUT)"
+          rm $GOFMT_OUT
+          if [ -n "$NEED_FMT" ]; then
+            echo "Go files not formatted:"
+            echo "$NEED_FMT"
+            exit 1
+          fi
+        else
+          ( echo "$files" | tr \\n \\0 | xargs -0 "$bin" $args >&2 ; times ) | ( read _ _; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+        fi
+        ;;
+      Java)
+          # Setting JAVA_RUNFILES to work around https://github.com/bazelbuild/bazel/issues/12348
+          ( echo "$files" | tr \\n \\0 | JAVA_RUNFILES="${RUNFILES_MANIFEST_FILE%_manifest}" xargs -0 "$bin" $args >&2 ; times ) | (read _ _ ;   read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+        ;;
+      Swift)
+        # for any linter that must be silenced
+        ( 2>/dev/null ; echo "$files" | tr \\n \\0 | xargs -0 "$bin" $args >&2 ; times ) | ( read _ _ ; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+        ;;
+      *)
+        ( echo "$files" | tr \\n \\0 | xargs -0 "$bin" $args >&2 ; times ) | ( read _ _ ; read tuser tsys; echo "Formatted ${lang} in ${tuser}" )
+        ;;
+    esac
   fi
-fi
-
-files=$(ls-files C++ $@)
-bin=$(rlocation {{clang-format}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  echo "Formatting C/C++ with clang-format..."
-  echo "$files" | tr \\n \\0 | xargs -0 $bin $clangformatmode
-fi
-
-files=$(ls-files Shell $@)
-bin=$(rlocation {{shfmt}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  echo "Formatting Shell with shfmt..."
-  echo "$files" | tr \\n \\0 | xargs -0 $bin $shfmtmode
-fi
+}
 
-files=$(ls-files Swift $@)
-bin=$(rlocation {{swiftformat}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  # swiftformat itself prints Running SwiftFormat...
-  echo "$files" | tr \\n \\0 | xargs -0 $bin $swiftmode
-fi
+# Run each supplied formatter over the files it owns
 
-files=$(ls-files 'Protocol Buffer' $@)
-bin=$(rlocation {{buf}})
-if [ -n "$files" ] && [ -n "$bin" ]; then
-  echo "Formatting Protobuf with buf..."
-  for file in $files; do
-    $bin $bufmode $file
-  done
-fi
+(
+  trap "kill 0" SIGINT
```

Comment:
could we break this up into two PRs please?

Let's land the refactoring to introduce the `run-lints` function first, but not run them in parallel yet. That way you don't get stuck with any messy rebases while we figure out how to buffer the output properly.